### PR TITLE
Result Types for Union

### PR DESCRIPTION
**Pull request by wboettge**
Currently, babelfish gives an error when UNION is used with fixed-length types of different lengths or null. This change adds hooks to select_common_type and select_common_typmod. This allows CHAR, NCHAR, and BINARY types to have the correct output type and length with UNION.

### Issues Resolved
[BABEL-3392](https://jira.rds.a2z.com/browse/BABEL-3392), [BABEL-3348](https://jira.rds.a2z.com/browse/BABEL-3348)

### Test Scenarios Covered ###
* **Use case based -**
Checks output types with combinations of the modified types. Verifies the output lengths are expected. Checks with UNION, UNION ALL, INTERSECT, and EXCEPT.

* **Boundary conditions -**
Checks with multiple nulls, multiple unions, with joins and subqueries. 

* **Arbitrary inputs -**
Validates multibyte chars are output correctly.

* **Negative test cases -**
Tests with TEXT, VARCHAR, and NVARCHAR to ensure behavior is unchanged

* **Minor version upgrade tests -**
Creates views in prepare step which contain error cases

* **Major version upgrade tests -**
Creates views in prepare step which contain error cases

* **Performance tests -**
N/A

* **Tooling impact -**
N/A

* **Client tests -**
N/A

### Check List
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: contrib/babelfishpg_tsql/src/pltsql_coerce.c [hunk 1]
```diff
@@ -1084,6 +1086,94 @@ tsql_coerce_string_literal_hook(ParseCallbackState *pcbstate, Oid targetTypeId,
...
+
+static bool
+expr_is_null_or_unknown(Node *expr)
+{
+	Oid type = exprType(expr);
+	return type == UNKNOWNOID || (
+		IsA(expr, Const) && (type == TEXTOID || ((Const*)expr)->constisnull));
+}
+
+static Oid
```
Comment (forestkeeper) [inline contrib/babelfishpg_tsql/src/pltsql_coerce.c:1124]: The return of this func
if it's invalidOid, it'll just follow the normal PG select_common_type
We should leave a comment here

**Review by forestkeeper — APPROVED** [review]

File: contrib/babelfishpg_tsql/src/pltsql_coerce.c [hunk 2]
```diff
@@ -1084,6 +1086,112 @@ tsql_coerce_string_literal_hook(ParseCallbackState *pcbstate, Oid targetTypeId,
...
+
+/* 
+ * Checks if the expression represents NULL or a string literal
+ */
+static bool
+expr_is_null_or_unknown(Node *expr)
+{
+	Oid type = exprType(expr);
+	return IsA(expr, Const) && (
+		type == UNKNOWNOID || type == TEXTOID || ((Const*)expr)->constisnull);
```
Comment (Deepesh125) [inline contrib/babelfishpg_tsql/src/pltsql_coerce.c:1113]: If NULL literal is supplied then how can type will be TEXTOID? In what cases?

File: contrib/babelfishpg_tsql/src/pltsql_coerce.c [hunk 3]
```diff
@@ -1084,6 +1086,112 @@ tsql_coerce_string_literal_hook(ParseCallbackState *pcbstate, Oid targetTypeId,
...
+/* 
+ * When we must merge types together (i.e. UNION), if all types are
+ * null, literals, or [n][var]char types, then return the correct
+ * output type based on TSQL's precedence rules
+ * 
+ * If InvalidOid is returned, engine will handle finding a common type
+ * as usual
+ */
+static Oid
+tsql_select_common_type_hook(ParseState *pstate, List *exprs, const char *context,
```
Comment (Deepesh125) [inline contrib/babelfishpg_tsql/src/pltsql_coerce.c:1125]: IMHO, just to be on safe side, we should call this hook only for UNION path as select_common_type is being used widely and we may break other operations ie, ISNULL.

Have we checked if there is any adverse impact on such operations due to this changes?

File: test/JDBC/input/BABEL-3392-vu-verify.sql [hunk 4]
```diff
@@ -0,0 +1,1027 @@
...
+
+-- SELECT CASE a
+--     WHEN 1 THEN NULL
+--     WHEN 2 THEN CAST('char' AS CHAR(10))
+-- END
+-- FROM babel3392_case
+-- GO
+
+-- DROP TABLE babel3392_case;
+-- GO
```
Comment (Deepesh125) [inline test/JDBC/input/BABEL-3392-vu-verify.sql:418]: why are these tests commented out?

File: contrib/babelfishpg_tsql/src/pltsql_coerce.c (hunk 3 above)
Comment (Deepesh125) [reply contrib/babelfishpg_tsql/src/pltsql_coerce.c:1125]: Also, our test cases only revolve around set operations, whereas this changes may affect other operations as wel as I said, ISNULL, case-when, MERGE join, etc

**Pull request merged**